.github: Add pytest workflow

underyx opened the pull request.

underyx committed:
.github: Add pytest workflow

underyx committed:
README: Rename 'generic' to 'any environment'

underyx commented on .github/workflows/test.yml:
```diff
@@ -3,11 +3,55 @@ name: Tests
+      - name: Install dependencies
+        run: |
+          poetry install
+          poetry add semgrep --no-interaction
```
1. the unpinned version bothers me
2. not sure if installing with poetry or with pipx (or pip) is better

brendongo replied:
Why not just pin it? We update semgrep version manually anyway. Just add the it should also be bumped here?

nbrahms replied:
I think you'd want to update the bump script in that case, as well

underyx replied:
No bump script here unfortunately, and wanted to avoid adding more manual work :(

nbrahms commented on tests/conftest.py:
```diff
@@ -9,10 +9,7 @@
-    "tool.driver.semanticVersion",
-    "results.extra.metavars.*.unique_id.md5sum",
-]
+MASKED_KEYS: List[str] = []
```
I don't understand this change...

If we're removing the functionality, shouldn't we just delete the constant and its use?

underyx replied:
Yeah, I wanted to add snapshot tests soon and wanted to also make it obvious that the existing elements were not used.

nbrahms commented on tests/conftest.py:
```diff
@@ -39,8 +36,13 @@ def _mark_masked(obj, path_items):
+        output = json.loads(output_json)
+    except json.JSONDecodeError:
+        raise ValueError(
+            f"Instead of JSON, output was:\n--- output start ---\n{output_json}\n--- output end ---"
```
```suggestion
            f"Instead of JSON, Semgrep output was:\n--- output start ---\n{output_json}\n--- output end ---"
```

nbrahms replied:
probably also use printing the semgrep stderr in this case somehow?

underyx replied:
it's printed above and pytest shows captured print output when a test fails

nbrahms commented on tests/conftest.py:
```diff
@@ -70,13 +70,15 @@ def _run_semgrep_agent(
 
     process = subprocess.run(
-        ["python3", "-m", "semgrep_agent", *options],
+        ["semgrep-agent", *options],
```
I think this does not work in some environments...

chmccreery replied:
can verify ^

underyx committed:
Address reviews

- pytest workflow installs a pinned semgrep version
- also, it installs it with pipx
- removed unused JSON key masking in tests
- tests invoke agent via python -m again

underyx committed:
.github: Cache whole poetry venv

underyx commented:
@brendongo can you re-review the part where we install semgrep for the pytest job?

brendongo commented on .github/workflows/test.yml:
```diff
@@ -3,11 +3,72 @@ name: Tests
+        run: |
+          # tests should simulate CI environment iff they need one
+          unset CI
+          unset "${!GITHUB_@}"
```
Does this unset all GITHUB_ env variables?

underyx replied:
yep
